The job **`x86_64-gnu-llvm-19`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/15193655283/job/42732484862) [(plain)](https://github.com/rust-lang/rust/commit/5b7da7933663c8dbaf5d0c13b8b4f36f40f6faf4/checks/42732484862/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#18 exporting to docker image format
#18 sending tarball 20.6s done
#18 DONE 29.0s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-19]
[CI_JOB_NAME=x86_64-gnu-llvm-19]
debug: `DISABLE_CI_RUSTC_IF_INCOMPATIBLE` configured.
---
sccache: Listening on address 127.0.0.1:4226
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-19', '--enable-llvm-link-shared', '--set', 'rust.randomize-layout=true', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'build.print-step-timings', '--enable-verbose-tests', '--set', 'build.metrics', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--set', 'rust.lld=false', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--set', 'gcc.download-ci-gcc=true', '--enable-new-symbol-mangling']
configure: build.build          := x86_64-unknown-linux-gnu
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-19/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.randomize-layout := True
configure: rust.thin-lto-import-instr-limit := 10
---
failures:

---- [codegen] tests/codegen/array-equality.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-19/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll" "/checkout/tests/codegen/array-equality.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen/array-equality.rs:9:17: error: CHECK-NEXT: is not on the line after the previous match
 // CHECK-NEXT: %2 = icmp eq i48 %0, %1
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:13:2: note: 'next' match was here
 %2 = icmp eq i48 %0, %1
 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:10:7: note: previous match ended here
start:
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:11:1: note: non-matching line after previous match is here
 %.sroa.03 = alloca i8, align 1
^
/checkout/tests/codegen/array-equality.rs:21:17: error: CHECK-NEXT: is not on the line after the previous match
 // CHECK-NEXT: ret
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:33:2: note: 'next' match was here
 ret i1 %_0
 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:28:18: note: previous match ended here
 %2 = icmp eq i48 %0, %1
                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:29:1: note: non-matching line after previous match is here
 store i1 %2, ptr %.sroa.0, align 1
^
/checkout/tests/codegen/array-equality.rs:31:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: ret i1 %[[EQ]]
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:42:27: note: scanning from here
 %0 = icmp eq i32 %bcmp, 0
                          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:42:27: note: with "EQ" equal to "0"
 %0 = icmp eq i32 %bcmp, 0
                          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:47:2: note: possible intended match here
 ret i1 %_0
 ^
/checkout/tests/codegen/array-equality.rs:41:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: ret i1 %[[EQ]]
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:56:27: note: scanning from here
 %0 = icmp eq i32 %bcmp, 0
                          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:56:27: note: with "EQ" equal to "0"
 %0 = icmp eq i32 %bcmp, 0
                          ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:61:2: note: possible intended match here
 ret i1 %_0
 ^
/checkout/tests/codegen/array-equality.rs:49:17: error: CHECK-NEXT: is not on the line after the previous match
 // CHECK-NEXT: %[[EQ:.+]] = icmp eq i64 %0, %1
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:69:2: note: 'next' match was here
 %2 = icmp eq i64 %0, %1
 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:66:7: note: previous match ended here
start:
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:67:1: note: non-matching line after previous match is here
 %.sroa.02 = alloca i8, align 1
^
/checkout/tests/codegen/array-equality.rs:58:17: error: CHECK-NEXT: is not on the line after the previous match
 // CHECK-NEXT: %[[EQ:.+]] = icmp eq i48 %0, 0
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:82:2: note: 'next' match was here
 %1 = icmp eq i48 %0, 0
 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:79:7: note: previous match ended here
start:
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:80:1: note: non-matching line after previous match is here
 %.sroa.01 = alloca i8, align 1
^
/checkout/tests/codegen/array-equality.rs:67:17: error: CHECK-NEXT: is not on the line after the previous match
 // CHECK-NEXT: %[[EQ:.+]] = icmp eq i40 %0, 0
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:95:2: note: 'next' match was here
 %1 = icmp eq i40 %0, 0
 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:92:7: note: previous match ended here
start:
      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:93:1: note: non-matching line after previous match is here
 %.sroa.01 = alloca i8, align 1
^
/checkout/tests/codegen/array-equality.rs:78:12: error: CHECK: expected string not found in input
 // CHECK: ret i1 %[[EQ]]
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:109:24: note: scanning from here
 %1 = icmp eq i72 %0, 0
                       ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:109:24: note: with "EQ" equal to "1"
 %1 = icmp eq i72 %0, 0
                       ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:114:2: note: possible intended match here
 ret i1 %_0
 ^
/checkout/tests/codegen/array-equality.rs:88:17: error: CHECK-NEXT: expected string not found in input
 // CHECK-NEXT: ret i1 %[[EQ]]
                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:123:25: note: scanning from here
 %1 = icmp eq i128 %0, 0
                        ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:123:25: note: with "EQ" equal to "1"
 %1 = icmp eq i128 %0, 0
                        ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:128:2: note: possible intended match here
 ret i1 %_0
 ^
/checkout/tests/codegen/array-equality.rs:96:16: error: CHECK-NOT: excluded string found in input
 // CHECK-NOT: alloca
               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll:134:13: note: found here
 %.sroa.0 = alloca i8, align 1
            ^~~~~~

Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/array-equality/array-equality.ll
Check file: /checkout/tests/codegen/array-equality.rs

-dump-input=help explains the following input dump.

Input was:
<<<<<<
            1: ; ModuleID = 'array_equality.e4333bdc97a88f9b-cgu.0' 
            2: source_filename = "array_equality.e4333bdc97a88f9b-cgu.0" 
            3: target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-i128:128-f80:128-n8:16:32:64-S128" 
            4: target triple = "x86_64-unknown-linux-gnu" 
            5:  
            6: @alloc_2f59a1ca6735fbffa4a0c2487d50fe27 = private unnamed_addr constant [2468 x i8] zeroinitializer, align 2 
            7:  
            8: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
            9: define noundef zeroext i1 @array_eq_value(i48 %0, i48 %1) unnamed_addr #0 { 
           10: start: 
           11:  %.sroa.03 = alloca i8, align 1 
           12:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.03) 
           13:  %2 = icmp eq i48 %0, %1 
next:9          !~~~~~~~~~~~~~~~~~~~~~~  error: match on wrong line
           14:  store i1 %2, ptr %.sroa.03, align 1 
           15:  %.sroa.03.0..sroa.03.0..sroa.03.0..sroa.03.0..sroa.03.0. = load i8, ptr %.sroa.03, align 1, !range !3, !noundef !4 
           16:  %_0 = trunc nuw i8 %.sroa.03.0..sroa.03.0..sroa.03.0..sroa.03.0..sroa.03.0. to i1 
           17:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.03) 
           18:  ret i1 %_0 
           19: } 
           20:  
           21: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable 
           22: define noundef zeroext i1 @array_eq_ref(ptr noalias nocapture noundef readonly align 2 dereferenceable(6) %a, ptr noalias nocapture noundef readonly align 2 dereferenceable(6) %b) unnamed_addr #1 { 
           23: start: 
           24:  %.sroa.0 = alloca i8, align 1 
           25:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.0) 
           26:  %0 = load i48, ptr %a, align 2 
           27:  %1 = load i48, ptr %b, align 2 
           28:  %2 = icmp eq i48 %0, %1 
           29:  store i1 %2, ptr %.sroa.0, align 1 
           30:  %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. = load i8, ptr %.sroa.0, align 1, !range !3, !noundef !4 
           31:  %_0 = trunc nuw i8 %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. to i1 
           32:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.0) 
           33:  ret i1 %_0 
next:21         !~~         error: match on wrong line
           34: } 
           35:  
           36: ; Function Attrs: mustprogress nofree nounwind nonlazybind willreturn memory(argmem: read) uwtable 
           37: define noundef zeroext i1 @array_eq_value_still_passed_by_pointer(ptr noalias nocapture noundef readonly align 2 dereferenceable(18) %a, ptr noalias nocapture noundef readonly align 2 dereferenceable(18) %b) unnamed_addr #2 { 
           38: start: 
           39:  %.sroa.0 = alloca i8, align 1 
           40:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.0) 
           41:  %bcmp = tail call i32 @bcmp(ptr noundef nonnull dereferenceable(18) %a, ptr noundef nonnull dereferenceable(18) %b, i64 18) 
           42:  %0 = icmp eq i32 %bcmp, 0 
next:31'0                                X error: no match found
next:31'1                                  with "EQ" equal to "0"
           43:  store i1 %0, ptr %.sroa.0, align 1 
next:31'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           44:  %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. = load i8, ptr %.sroa.0, align 1, !range !3, !noundef !4 
next:31'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           45:  %_0 = trunc nuw i8 %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. to i1 
next:31'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           46:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.0) 
next:31'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           47:  ret i1 %_0 
next:31'0      ~~~~~~~~~~~~
next:31'2       ?           possible intended match
           48: } 
next:31'0      ~~
           49:  
next:31'0      ~
           50: ; Function Attrs: mustprogress nofree nounwind nonlazybind willreturn memory(argmem: read) uwtable 
next:31'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           51: define noundef zeroext i1 @array_eq_long(ptr noalias nocapture noundef readonly align 2 dereferenceable(2468) %a, ptr noalias nocapture noundef readonly align 2 dereferenceable(2468) %b) unnamed_addr #2 { 
next:31'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           52: start: 
           53:  %.sroa.0 = alloca i8, align 1 
           54:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.0) 
           55:  %bcmp = tail call i32 @bcmp(ptr noundef nonnull dereferenceable(2468) %a, ptr noundef nonnull dereferenceable(2468) %b, i64 2468) 
           56:  %0 = icmp eq i32 %bcmp, 0 
next:41'0                                X error: no match found
next:41'1                                  with "EQ" equal to "0"
           57:  store i1 %0, ptr %.sroa.0, align 1 
next:41'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           58:  %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. = load i8, ptr %.sroa.0, align 1, !range !3, !noundef !4 
next:41'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           59:  %_0 = trunc nuw i8 %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. to i1 
next:41'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           60:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.0) 
next:41'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           61:  ret i1 %_0 
next:41'0      ~~~~~~~~~~~~
next:41'2       ?           possible intended match
           62: } 
next:41'0      ~~
           63:  
next:41'0      ~
           64: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
next:41'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           65: define noundef zeroext i1 @array_char_eq(i64 %0, i64 %1) unnamed_addr #0 { 
next:41'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
           66: start: 
           67:  %.sroa.02 = alloca i8, align 1 
           68:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.02) 
           69:  %2 = icmp eq i64 %0, %1 
next:49         !~~~~~~~~~~~~~~~~~~~~~~  error: match on wrong line
           70:  store i1 %2, ptr %.sroa.02, align 1 
           71:  %.sroa.02.0..sroa.02.0..sroa.02.0..sroa.02.0..sroa.02.0. = load i8, ptr %.sroa.02, align 1, !range !3, !noundef !4 
           72:  %_0 = trunc nuw i8 %.sroa.02.0..sroa.02.0..sroa.02.0..sroa.02.0..sroa.02.0. to i1 
           73:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.02) 
           74:  ret i1 %_0 
           75: } 
           76:  
           77: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
           78: define noundef zeroext i1 @array_eq_zero_short(i48 %0) unnamed_addr #0 { 
           79: start: 
           80:  %.sroa.01 = alloca i8, align 1 
           81:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.01) 
           82:  %1 = icmp eq i48 %0, 0 
next:58         !~~~~~~~~~~~~~~~~~~~~~  error: match on wrong line
           83:  store i1 %1, ptr %.sroa.01, align 1 
           84:  %.sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0. = load i8, ptr %.sroa.01, align 1, !range !3, !noundef !4 
           85:  %_0 = trunc nuw i8 %.sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0. to i1 
           86:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.01) 
           87:  ret i1 %_0 
           88: } 
           89:  
           90: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable 
           91: define noundef zeroext i1 @array_eq_none_short(i40 %0) unnamed_addr #0 { 
           92: start: 
           93:  %.sroa.01 = alloca i8, align 1 
           94:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.01) 
           95:  %1 = icmp eq i40 %0, 0 
next:67         !~~~~~~~~~~~~~~~~~~~~~  error: match on wrong line
           96:  store i1 %1, ptr %.sroa.01, align 1 
           97:  %.sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0. = load i8, ptr %.sroa.01, align 1, !range !3, !noundef !4 
           98:  %_0 = trunc nuw i8 %.sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0..sroa.01.0. to i1 
           99:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.01) 
          100:  ret i1 %_0 
          101: } 
          102:  
          103: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable 
          104: define noundef zeroext i1 @array_eq_zero_nested(ptr noalias nocapture noundef readonly align 1 dereferenceable(9) %x) unnamed_addr #1 { 
          105: start: 
          106:  %.sroa.0 = alloca i8, align 1 
          107:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.0) 
          108:  %0 = load i72, ptr %x, align 1 
          109:  %1 = icmp eq i72 %0, 0 
check:78'0                            X error: no match found
check:78'1                              with "EQ" equal to "1"
          110:  store i1 %1, ptr %.sroa.0, align 1 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          111:  %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. = load i8, ptr %.sroa.0, align 1, !range !3, !noundef !4 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          112:  %_0 = trunc nuw i8 %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. to i1 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          113:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.0) 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          114:  ret i1 %_0 
check:78'0     ~~~~~~~~~~~~
check:78'2      ?           possible intended match
          115: } 
check:78'0     ~~
          116:  
check:78'0     ~
          117: ; Function Attrs: mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          118: define noundef zeroext i1 @array_eq_zero_mid(ptr noalias nocapture noundef readonly align 2 dereferenceable(16) %x) unnamed_addr #1 { 
check:78'0     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          119: start: 
          120:  %.sroa.0 = alloca i8, align 1 
          121:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.0) 
          122:  %0 = load i128, ptr %x, align 2 
          123:  %1 = icmp eq i128 %0, 0 
next:88'0                              X error: no match found
next:88'1                                with "EQ" equal to "1"
          124:  store i1 %1, ptr %.sroa.0, align 1 
next:88'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          125:  %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. = load i8, ptr %.sroa.0, align 1, !range !3, !noundef !4 
next:88'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          126:  %_0 = trunc nuw i8 %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. to i1 
next:88'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          127:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.0) 
next:88'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          128:  ret i1 %_0 
next:88'0      ~~~~~~~~~~~~
next:88'2       ?           possible intended match
          129: } 
next:88'0      ~~
          130:  
next:88'0      ~
          131: ; Function Attrs: mustprogress nofree nounwind nonlazybind willreturn memory(argmem: read) uwtable 
next:88'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          132: define noundef zeroext i1 @array_eq_zero_long(ptr noalias nocapture noundef readonly align 2 dereferenceable(2468) %x) unnamed_addr #2 { 
next:88'0      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
          133: start: 
          134:  %.sroa.0 = alloca i8, align 1 
not:96                     !~~~~~              error: no match expected
          135:  call void @llvm.lifetime.start.p0(i64 1, ptr nonnull %.sroa.0) 
          136:  %bcmp = tail call i32 @bcmp(ptr noundef nonnull dereferenceable(2468) %x, ptr noundef nonnull dereferenceable(2468) @alloc_2f59a1ca6735fbffa4a0c2487d50fe27, i64 2468) 
          137:  %0 = icmp eq i32 %bcmp, 0 
          138:  store i1 %0, ptr %.sroa.0, align 1 
          139:  %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. = load i8, ptr %.sroa.0, align 1, !range !3, !noundef !4 
          140:  %_0 = trunc nuw i8 %.sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0..sroa.0.0. to i1 
          141:  call void @llvm.lifetime.end.p0(i64 1, ptr nonnull %.sroa.0) 
          142:  ret i1 %_0 
          143: } 
          144:  
          145: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
          146: declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #3 
          147:  
          148: ; Function Attrs: mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) 
          149: declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #3 
          150:  
          151: ; Function Attrs: nofree nounwind nonlazybind willreturn memory(argmem: read) 
          152: declare i32 @bcmp(ptr nocapture, ptr nocapture, i64) local_unnamed_addr #4 
          153:  
          154: attributes #0 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(none) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
          155: attributes #1 = { mustprogress nofree norecurse nosync nounwind nonlazybind willreturn memory(argmem: read) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
          156: attributes #2 = { mustprogress nofree nounwind nonlazybind willreturn memory(argmem: read) uwtable "probe-stack"="inline-asm" "target-cpu"="x86-64" } 
          157: attributes #3 = { mustprogress nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) } 
          158: attributes #4 = { nofree nounwind nonlazybind willreturn memory(argmem: read) } 
          159:  
          160: !llvm.module.flags = !{!0, !1} 
          161: !llvm.ident = !{!2} 
          162:  
          163: !0 = !{i32 8, !"PIC Level", i32 2} 
          164: !1 = !{i32 2, !"RtLibUseGOT", i32 1} 
          165: !2 = !{!"rustc version 1.89.0-nightly (d78369d2b 2025-05-22)"} 
          166: !3 = !{i8 0, i8 2} 
          167: !4 = !{} 
>>>>>>
------------------------------------------


---- [codegen] tests/codegen/is_val_statically_known.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-19/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll" "/checkout/tests/codegen/is_val_statically_known.rs" "--check-prefix=CHECK" "--allow-unused-prefixes" "--dump-input-context" "100"
stdout: none
--- stderr -------------------------------
/checkout/tests/codegen/is_val_statically_known.rs:21:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 1
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:7:47: note: scanning from here
define noundef range(i32 0, 2) i32 @_u32_true() unnamed_addr #0 {
                                              ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:15:2: note: possible intended match here
 ret i32 %_1
 ^
/checkout/tests/codegen/is_val_statically_known.rs:28:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 0
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:19:48: note: scanning from here
define noundef range(i32 0, 2) i32 @_u32_false(i32 noundef %a) unnamed_addr #0 {
                                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:27:2: note: possible intended match here
 ret i32 %_2
 ^
/checkout/tests/codegen/is_val_statically_known.rs:40:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 3
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:31:48: note: scanning from here
define noundef range(i32 2, 4) i32 @_bool_true() unnamed_addr #0 {
                                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:40:2: note: possible intended match here
 ret i32 %.
 ^
/checkout/tests/codegen/is_val_statically_known.rs:47:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 2
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:44:49: note: scanning from here
define noundef range(i32 2, 4) i32 @_bool_false(i1 noundef zeroext %b) unnamed_addr #0 {
                                                ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:53:2: note: possible intended match here
 ret i32 %.
 ^
/checkout/tests/codegen/is_val_statically_known.rs:59:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 4
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:57:50: note: scanning from here
define noundef range(i32 4, 6) i32 @_iref_borrow() unnamed_addr #0 {
                                                 ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:66:2: note: possible intended match here
 ret i32 %.
 ^
/checkout/tests/codegen/is_val_statically_known.rs:66:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 4
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:70:47: note: scanning from here
define noundef range(i32 4, 6) i32 @_iref_arg(ptr noalias noundef readonly align 1 dereferenceable(1) %a) unnamed_addr #0 {
                                              ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:79:2: note: possible intended match here
 ret i32 %.
 ^
/checkout/tests/codegen/is_val_statically_known.rs:78:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 6
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:83:55: note: scanning from here
define noundef range(i32 6, 8) i32 @_slice_ref_borrow() unnamed_addr #0 {
                                                      ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:92:2: note: possible intended match here
 ret i32 %.
 ^
/checkout/tests/codegen/is_val_statically_known.rs:85:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 6
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:96:52: note: scanning from here
define noundef range(i32 6, 8) i32 @_slice_ref_arg(ptr noalias nocapture noundef nonnull readonly align 1 %a.0, i64 noundef %a.1) unnamed_addr #0 {
                                                   ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:105:2: note: possible intended match here
 ret i32 %.
 ^
/checkout/tests/codegen/is_val_statically_known.rs:97:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 1
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:109:47: note: scanning from here
define noundef range(i32 0, 2) i32 @_f16_true() unnamed_addr #0 {
                                              ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:117:2: note: possible intended match here
 ret i32 %_1
 ^
/checkout/tests/codegen/is_val_statically_known.rs:104:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 0
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:121:48: note: scanning from here
define noundef range(i32 0, 2) i32 @_f16_false(half noundef %a) unnamed_addr #0 {
                                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:129:2: note: possible intended match here
 ret i32 %_2
 ^
/checkout/tests/codegen/is_val_statically_known.rs:116:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 1
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:133:47: note: scanning from here
define noundef range(i32 0, 2) i32 @_f32_true() unnamed_addr #0 {
                                              ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:141:2: note: possible intended match here
 ret i32 %_1
 ^
/checkout/tests/codegen/is_val_statically_known.rs:123:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 0
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:145:48: note: scanning from here
define noundef range(i32 0, 2) i32 @_f32_false(float noundef %a) unnamed_addr #0 {
                                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:153:2: note: possible intended match here
 ret i32 %_2
 ^
/checkout/tests/codegen/is_val_statically_known.rs:135:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 1
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:157:47: note: scanning from here
define noundef range(i32 0, 2) i32 @_f64_true() unnamed_addr #0 {
                                              ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:165:2: note: possible intended match here
 ret i32 %_1
 ^
/checkout/tests/codegen/is_val_statically_known.rs:142:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 0
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:169:48: note: scanning from here
define noundef range(i32 0, 2) i32 @_f64_false(double noundef %a) unnamed_addr #0 {
                                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:177:2: note: possible intended match here
 ret i32 %_2
 ^
/checkout/tests/codegen/is_val_statically_known.rs:154:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 1
           ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:181:48: note: scanning from here
define noundef range(i32 0, 2) i32 @_f128_true() unnamed_addr #0 {
                                               ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/is_val_statically_known/is_val_statically_known.ll:189:2: note: possible intended match here
 ret i32 %_1
 ^
/checkout/tests/codegen/is_val_statically_known.rs:161:12: error: CHECK: expected string not found in input
 // CHECK: ret i32 0
```

</details>

--- a/compiler/rustc_codegen_ssa/src/mir/block.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/block.rs
@@ -8,7 +8,7 @@ use rustc_hir::lang_items::LangItem;
 use rustc_middle::mir::{self, AssertKind, InlineAsmMacro, SwitchTargets, UnwindTerminateReason};
 use rustc_middle::ty::layout::{HasTyCtxt, LayoutOf, ValidityRequirement};
 use rustc_middle::ty::print::{with_no_trimmed_paths, with_no_visible_paths};
-use rustc_middle::ty::{self, Instance, Ty};
+use rustc_middle::ty::{self, Instance, List, Ty};
 use rustc_middle::{bug, span_bug};
 use rustc_session::config::OptLevel;
 use rustc_span::source_map::Spanned;
@@ -827,7 +827,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         helper: &TerminatorCodegenHelper<'tcx>,
         bx: &mut Bx,
         intrinsic: ty::IntrinsicDef,
-        instance: Option<Instance<'tcx>>,
+        instance: Instance<'tcx>,
         source_info: mir::SourceInfo,
         target: Option<mir::BasicBlock>,
         unwind: mir::UnwindAction,
@@ -837,7 +837,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         // These are intrinsics that compile to panics so that we can get a message
         // which mentions the offending type, even from a const context.
         if let Some(requirement) = ValidityRequirement::from_intrinsic(intrinsic.name) {
-            let ty = instance.unwrap().args.type_at(0);
+            let ty = instance.args.type_at(0);
 
             let do_panic = !bx
                 .tcx()
@@ -910,35 +910,116 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         let callee = self.codegen_operand(bx, func);
 
         let (instance, mut llfn) = match *callee.layout.ty.kind() {
-            ty::FnDef(def_id, args) => (
-                Some(ty::Instance::expect_resolve(
+            ty::FnDef(def_id, generic_args) => {
+                let instance = ty::Instance::expect_resolve(
                     bx.tcx(),
                     bx.typing_env(),
                     def_id,
-                    args,
+                    generic_args,
                     fn_span,
-                )),
-                None,
-            ),
+                );
+
+                let instance = match instance.def {
+                    // We don't need AsyncDropGlueCtorShim here because it is not `noop func`,
+                    // it is `func returning noop future`
+                    ty::InstanceKind::DropGlue(_, None) => {
+                        // Empty drop glue; a no-op.
+                        let target = target.unwrap();
+                        return helper.funclet_br(self, bx, target, mergeable_succ);
+                    }
+                    ty::InstanceKind::Intrinsic(def_id) => {
+                        let intrinsic = bx.tcx().intrinsic(def_id).unwrap();
+                        if let Some(merging_succ) = self.codegen_panic_intrinsic(
+                            &helper,
+                            bx,
+                            intrinsic,
+                            instance,
+                            source_info,
+                            target,
+                            unwind,
+                            mergeable_succ,
+                        ) {
+                            return merging_succ;
+                        }
+
+                        let fn_abi = bx.fn_abi_of_instance(instance, List::empty());
+
+                        let mut llargs = Vec::with_capacity(1);
+                        let ret_dest = self.make_return_dest(
+                            bx,
+                            destination,
+                            &fn_abi.ret,
+                            &mut llargs,
+                            Some(intrinsic),
+                        );
+                        let dest = match ret_dest {
+                            _ if fn_abi.ret.is_indirect() => llargs[0],
+                            ReturnDest::Nothing => bx.const_undef(bx.type_ptr()),
+                            ReturnDest::IndirectOperand(dst, _) | ReturnDest::Store(dst) => {
+                                dst.val.llval
+                            }
+                            ReturnDest::DirectOperand(_) => {
+                                bug!("Cannot use direct operand with an intrinsic call")
+                            }
+                        };
+
+                        let args: Vec<_> =
+                            args.iter().map(|arg| self.codegen_operand(bx, &arg.node)).collect();
+
+                        if matches!(intrinsic, ty::IntrinsicDef { name: sym::caller_location, .. })
+                        {
+                            let location = self.get_caller_location(
+                                bx,
+                                mir::SourceInfo { span: fn_span, ..source_info },
+                            );
+
+                            assert_eq!(llargs, []);
+                            if let ReturnDest::IndirectOperand(tmp, _) = ret_dest {
+                                location.val.store(bx, tmp);
+                            }
+                            self.store_return(bx, ret_dest, &fn_abi.ret, location.immediate());
+                            return helper.funclet_br(self, bx, target.unwrap(), mergeable_succ);
+                        }
+
+                        match Self::codegen_intrinsic_call(bx, instance, fn_abi, &args, dest, span)
+                        {
+                            Ok(()) => {
+                                if let ReturnDest::IndirectOperand(dst, _) = ret_dest {
+                                    self.store_return(bx, ret_dest, &fn_abi.ret, dst.val.llval);
+                                }
+
+                                return if let Some(target) = target {
+                                    helper.funclet_br(self, bx, target, mergeable_succ)
+                                } else {
+                                    bx.unreachable();
+                                    MergingSucc::False
+                                };
+                            }
+                            Err(instance) => {
+                                if intrinsic.must_be_overridden {
+                                    span_bug!(
+                                        span,
+                                        "intrinsic {} must be overridden by codegen backend, but isn't",
+                                        intrinsic.name,
+                                    );
+                                }
+                                instance
+                            }
+                        }
+                    }
+                    _ => instance,
+                };
+
+                (Some(instance), None)
+            }
             ty::FnPtr(..) => (None, Some(callee.immediate())),
             _ => bug!("{} is not callable", callee.layout.ty),
         };
-
-        let def = instance.map(|i| i.def);
-
-        // We don't need AsyncDropGlueCtorShim here because it is not `noop func`,
-        // it is `func returning noop future`
-        if let Some(ty::InstanceKind::DropGlue(_, None)) = def {
-            // Empty drop glue; a no-op.
-            let target = target.unwrap();
-            return helper.funclet_br(self, bx, target, mergeable_succ);
-        }
 
         // FIXME(eddyb) avoid computing this if possible, when `instance` is
         // available - right now `sig` is only needed for getting the `abi`
         // and figuring out how many extra args were passed to a C-variadic `fn`.
         let sig = callee.layout.ty.fn_sig(bx.tcx());
-        let abi = sig.abi();
 
         let extra_args = &args[sig.inputs().skip_binder().len()..];
         let extra_args = bx.tcx().mk_type_list_from_iter(extra_args.iter().map(|op_arg| {
@@ -954,83 +1035,6 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         // The arguments we'll be passing. Plus one to account for outptr, if used.
         let arg_count = fn_abi.args.len() + fn_abi.ret.is_indirect() as usize;
 
-        let instance = match def {
-            Some(ty::InstanceKind::Intrinsic(def_id)) => {
-                let intrinsic = bx.tcx().intrinsic(def_id).unwrap();
-                if let Some(merging_succ) = self.codegen_panic_intrinsic(
-                    &helper,
-                    bx,
-                    intrinsic,
-                    instance,
-                    source_info,
-                    target,
-                    unwind,
-                    mergeable_succ,
-                ) {
-                    return merging_succ;
-                }
-
-                let mut llargs = Vec::with_capacity(1);
-                let ret_dest = self.make_return_dest(
-                    bx,
-                    destination,
-                    &fn_abi.ret,
-                    &mut llargs,
-                    Some(intrinsic),
-                );
-                let dest = match ret_dest {
-                    _ if fn_abi.ret.is_indirect() => llargs[0],
-                    ReturnDest::Nothing => bx.const_undef(bx.type_ptr()),
-                    ReturnDest::IndirectOperand(dst, _) | ReturnDest::Store(dst) => dst.val.llval,
-                    ReturnDest::DirectOperand(_) => {
-                        bug!("Cannot use direct operand with an intrinsic call")
-                    }
-                };
-
-                let args: Vec<_> =
-                    args.iter().map(|arg| self.codegen_operand(bx, &arg.node)).collect();
-
-                if matches!(intrinsic, ty::IntrinsicDef { name: sym::caller_location, .. }) {
-                    let location = self
-                        .get_caller_location(bx, mir::SourceInfo { span: fn_span, ..source_info });
-
-                    assert_eq!(llargs, []);
-                    if let ReturnDest::IndirectOperand(tmp, _) = ret_dest {
-                        location.val.store(bx, tmp);
-                    }
-                    self.store_return(bx, ret_dest, &fn_abi.ret, location.immediate());
-                    return helper.funclet_br(self, bx, target.unwrap(), mergeable_succ);
-                }
-
-                let instance = *instance.as_ref().unwrap();
-                match Self::codegen_intrinsic_call(bx, instance, fn_abi, &args, dest, span) {
-                    Ok(()) => {
-                        if let ReturnDest::IndirectOperand(dst, _) = ret_dest {
-                            self.store_return(bx, ret_dest, &fn_abi.ret, dst.val.llval);
-                        }
-
-                        return if let Some(target) = target {
-                            helper.funclet_br(self, bx, target, mergeable_succ)
-                        } else {
-                            bx.unreachable();
-                            MergingSucc::False
-                        };
-                    }
-                    Err(instance) => {
-                        if intrinsic.must_be_overridden {
-                            span_bug!(
-                                span,
-                                "intrinsic {} must be overridden by codegen backend, but isn't",
-                                intrinsic.name,
-                            );
-                        }
-                        Some(instance)
-                    }
-                }
-            }
-            _ => instance,
-        };
-
         let mut llargs = Vec::with_capacity(arg_count);
 
         // We still need to call `make_return_dest` even if there's no `target`, since
@@ -1040,7 +1044,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         let destination = target.map(|target| (return_dest, target));
 
         // Split the rust-call tupled arguments off.
-        let (first_args, untuple) = if abi == ExternAbi::RustCall
+        let (first_args, untuple) = if sig.abi() == ExternAbi::RustCall
             && let Some((tup, args)) = args.split_last()
         {
             (args, Some(tup))
@@ -1055,7 +1059,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         'make_args: for (i, arg) in first_args.iter().enumerate() {
             let mut op = self.codegen_operand(bx, &arg.node);
 
-            if let (0, Some(ty::InstanceKind::Virtual(_, idx))) = (i, def) {
+            if let (0, Some(ty::InstanceKind::Virtual(_, idx))) = (i, instance.map(|i| i.def)) {
                 match op.val {
                     Pair(data_ptr, meta) => {
                         // In the case of Rc<Self>, we need to explicitly pass a

--- a/compiler/rustc_codegen_ssa/src/mir/block.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/block.rs
@@ -11,8 +11,8 @@ use rustc_middle::ty::print::{with_no_trimmed_paths, with_no_visible_paths};
 use rustc_middle::ty::{self, Instance, List, Ty};
 use rustc_middle::{bug, span_bug};
 use rustc_session::config::OptLevel;
+use rustc_span::Span;
 use rustc_span::source_map::Spanned;
-use rustc_span::{Span, sym};
 use rustc_target::callconv::{ArgAbi, FnAbi, PassMode};
 use tracing::{debug, info};
 
@@ -965,20 +965,13 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                         let args: Vec<_> =
                             args.iter().map(|arg| self.codegen_operand(bx, &arg.node)).collect();
 
-                        if matches!(intrinsic, ty::IntrinsicDef { name: sym::caller_location, .. })
-                        {
-                            let location = self.get_caller_location(bx, source_info);
-
-                            assert_eq!(llargs, []);
-                            if let ReturnDest::IndirectOperand(tmp, _) = ret_dest {
-                                location.val.store(bx, tmp);
-                            }
-                            self.store_return(bx, ret_dest, &fn_abi.ret, location.immediate());
-                            return helper.funclet_br(self, bx, target.unwrap(), mergeable_succ);
-                        }
-
-                        match Self::codegen_intrinsic_call(
-                            bx, instance, fn_abi, &args, dest, fn_span,
+                        match self.codegen_intrinsic_call(
+                            bx,
+                            instance,
+                            fn_abi,
+                            &args,
+                            dest,
+                            source_info,
                         ) {
                             Ok(()) => {
                                 if let ReturnDest::IndirectOperand(dst, _) = ret_dest {
@@ -1667,7 +1660,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         tuple.layout.fields.count()
     }
 
-    fn get_caller_location(
+    pub(super) fn get_caller_location(
         &mut self,
         bx: &mut Bx,
         source_info: mir::SourceInfo,

--- a/compiler/rustc_codegen_ssa/src/mir/block.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/block.rs
@@ -903,8 +903,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         fn_span: Span,
         mergeable_succ: bool,
     ) -> MergingSucc {
-        let source_info = terminator.source_info;
-        let span = source_info.span;
+        let source_info = mir::SourceInfo { span: fn_span, ..terminator.source_info };
 
         // Create the callee. This is a fn ptr or zero-sized and hence a kind of scalar.
         let callee = self.codegen_operand(bx, func);
@@ -968,10 +967,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
 
                         if matches!(intrinsic, ty::IntrinsicDef { name: sym::caller_location, .. })
                         {
-                            let location = self.get_caller_location(
-                                bx,
-                                mir::SourceInfo { span: fn_span, ..source_info },
-                            );
+                            let location = self.get_caller_location(bx, source_info);
 
                             assert_eq!(llargs, []);
                             if let ReturnDest::IndirectOperand(tmp, _) = ret_dest {
@@ -981,8 +977,9 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                             return helper.funclet_br(self, bx, target.unwrap(), mergeable_succ);
                         }
 
-                        match Self::codegen_intrinsic_call(bx, instance, fn_abi, &args, dest, span)
-                        {
+                        match Self::codegen_intrinsic_call(
+                            bx, instance, fn_abi, &args, dest, fn_span,
+                        ) {
                             Ok(()) => {
                                 if let ReturnDest::IndirectOperand(dst, _) = ret_dest {
                                     self.store_return(bx, ret_dest, &fn_abi.ret, dst.val.llval);
@@ -998,7 +995,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                             Err(instance) => {
                                 if intrinsic.must_be_overridden {
                                     span_bug!(
-                                        span,
+                                        fn_span,
                                         "intrinsic {} must be overridden by codegen backend, but isn't",
                                         intrinsic.name,
                                     );
@@ -1113,7 +1110,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                         // Make sure that we've actually unwrapped the rcvr down
                         // to a pointer or ref to `dyn* Trait`.
                         if !op.layout.ty.builtin_deref(true).unwrap().is_dyn_star() {
-                            span_bug!(span, "can't codegen a virtual call on {:#?}", op);
+                            span_bug!(fn_span, "can't codegen a virtual call on {:#?}", op);
                         }
                         let place = op.deref(bx.cx());
                         let data_place = place.project_field(bx, 0);
@@ -1129,7 +1126,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                         continue;
                     }
                     _ => {
-                        span_bug!(span, "can't codegen a virtual call on {:#?}", op);
+                        span_bug!(fn_span, "can't codegen a virtual call on {:#?}", op);
                     }
                 }
             }
@@ -1179,8 +1176,7 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
                 mir_args + 1,
                 "#[track_caller] fn's must have 1 more argument in their ABI than in their MIR: {instance:?} {fn_span:?} {fn_abi:?}",
             );
-            let location =
-                self.get_caller_location(bx, mir::SourceInfo { span: fn_span, ..source_info });
+            let location = self.get_caller_location(bx, source_info);
             debug!(
                 "codegen_call_terminator({:?}): location={:?} (fn_span {:?})",
                 terminator, location, fn_span
@@ -1199,9 +1195,9 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         let fn_ptr = match (instance, llfn) {
             (Some(instance), None) => bx.get_fn_addr(instance),
             (_, Some(llfn)) => llfn,
-            _ => span_bug!(span, "no instance or llfn for call"),
+            _ => span_bug!(fn_span, "no instance or llfn for call"),
         };
-        self.set_debug_loc(bx, mir::SourceInfo { span: fn_span, ..source_info });
+        self.set_debug_loc(bx, source_info);
         helper.do_call(
             self,
             bx,

--- a/compiler/rustc_codegen_ssa/src/mir/intrinsic.rs
+++ b/compiler/rustc_codegen_ssa/src/mir/intrinsic.rs
@@ -1,8 +1,9 @@
 use rustc_abi::WrappingRange;
+use rustc_middle::mir::SourceInfo;
 use rustc_middle::ty::{self, Ty, TyCtxt};
 use rustc_middle::{bug, span_bug};
 use rustc_session::config::OptLevel;
-use rustc_span::{Span, sym};
+use rustc_span::sym;
 use rustc_target::callconv::{FnAbi, PassMode};
 
 use super::FunctionCx;
@@ -52,13 +53,15 @@ fn memset_intrinsic<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>>(
 impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
     /// In the `Err` case, returns the instance that should be called instead.
     pub fn codegen_intrinsic_call(
+        &mut self,
         bx: &mut Bx,
         instance: ty::Instance<'tcx>,
         fn_abi: &FnAbi<'tcx, Ty<'tcx>>,
         args: &[OperandRef<'tcx, Bx::Value>],
         llresult: Bx::Value,
-        span: Span,
+        source_info: SourceInfo,
     ) -> Result<(), ty::Instance<'tcx>> {
+        let span = source_info.span;
         let callee_ty = instance.ty(bx.tcx(), bx.typing_env());
 
         let ty::FnDef(def_id, fn_args) = *callee_ty.kind() else {
@@ -102,6 +105,12 @@ impl<'a, 'tcx, Bx: BuilderMethods<'a, 'tcx>> FunctionCx<'a, 'tcx, Bx> {
         let llval = match name {
             sym::abort => {
                 bx.abort();
+                return Ok(());
+            }
+
+            sym::caller_location => {
+                let location = self.get_caller_location(bx, source_info);
+                location.val.store(bx, result);
                 return Ok(());
             }
 

